### PR TITLE
docker: 27.3.1 -> 27.4.0

### PR DESCRIPTION
**Pull request by teutat3s**
Built on x86_64-linux:
- `pkgs.docker_27`
- `pkgs.docker_27.passthru.tests`

Superseedes https://github.com/NixOS/nixpkgs/pull/363776

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/applications/virtualization/docker/default.nix [hunk 1]
```diff
@@ -45,6 +45,10 @@ rec {
         hash = runcHash;
       };
 
+      preBuild = ''
+        substituteInPlace Makefile --replace-fail "/bin/bash" "${bash}/bin/bash"
```
Comment (getchoo) [inline pkgs/applications/virtualization/docker/default.nix:49]:
```suggestion
        substituteInPlace Makefile --replace-quiet "/bin/bash" "${bash}/bin/bash"
```
This breaks builds on v25 and v26

Could also use --replace-warn. Up to you

**Comment by teutat3s** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 364247`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 16 packages built:</summary>
  <ul>
    <li>airlift</li>
    <li>airlift.dist</li>
    <li>charliecloud</li>
    <li>devcontainer</li>
    <li>docker (docker_27)</li>
    <li>docker-client</li>
    <li>docker-gc</li>
    <li>docker-sbom</li>
    <li>docker_25</li>
    <li>docker_26</li>
    <li>fn-cli</li>
    <li>pipework</li>
    <li>python311Packages.jupyter-repo2docker</li>
    <li>python311Packages.jupyter-repo2docker.dist</li>
    <li>python312Packages.jupyter-repo2docker</li>
    <li>python312Packages.jupyter-repo2docker.dist</li>
  </ul>
</details>

File: pkgs/applications/virtualization/docker/default.nix [hunk 2]
```diff
@@ -45,6 +45,10 @@ rec {
         hash = runcHash;
       };
 
+      preBuild = ''
+        substituteInPlace Makefile --replace-warn "/bin/bash" "${bash}/bin/bash"
```
Comment (getchoo) [inline pkgs/applications/virtualization/docker/default.nix:49]:
```suggestion
        substituteInPlace Makefile --replace-warn "/bin/bash" "${lib.getExe bash}"
```
Nitpick

**Review by getchoo — APPROVED** [review]
LGTM and everything builds. Feel free to ping in a couple days if there are no other reviews :)

## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 364247`

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 16 packages built:</summary>
  <ul>
    <li>airlift</li>
    <li>airlift.dist</li>
    <li>charliecloud</li>
    <li>devcontainer</li>
    <li>docker</li>
    <li>docker-client</li>
    <li>docker-gc</li>
    <li>docker-sbom</li>
    <li>docker_25</li>
    <li>docker_26</li>
    <li>fn-cli</li>
    <li>pipework</li>
    <li>python311Packages.jupyter-repo2docker</li>
    <li>python311Packages.jupyter-repo2docker.dist</li>
    <li>python312Packages.jupyter-repo2docker</li>
    <li>python312Packages.jupyter-repo2docker.dist</li>
  </ul>
</details>

**Review by getchoo — APPROVED** [review]
Carrying over my review

**Pull request merged**